File: client/src/boundary/interactor/StartGameInteractor.test.ts
import GameStorage from "services/api/GameStorage";
import { GameD2BConverter } from "./GameD2BConverter";
import { Board, Game } from "domain/index";
import { BoundaryBoard, BoundaryGame } from "boundary/model/index";
import { StartGameInteractor } from "./StartGameInteractor";
import { mock, MockProxy } from "jest-mock-extended";
import { of } from "rxjs";
import { GameState } from "utils/Constants";

describe("Start Game Interactor", () => {
  let gameStorage: MockProxy<GameStorage>;
  let gameD2BConverter: MockProxy<GameD2BConverter>;
  let startGameInteractor: StartGameInteractor;
  const gameId = "123";

  beforeEach(() => {
    gameStorage = mock<GameStorage>();
    gameD2BConverter = mock<GameD2BConverter>();
    startGameInteractor = new StartGameInteractor(
      gameStorage,
      gameD2BConverter
    );
  });

  it("creates a game and returns it", (done) => {
    const expectedGame: Game = new Game(
      gameId,
      GameState.IN_PROGRESS,
      mock<Board>(),
      25,
      0
    );
    const expectedBoundaryGame: BoundaryGame = new BoundaryGame(
      gameId,
      GameState.IN_PROGRESS,
      mock<BoundaryBoard>()
    );

    gameStorage.startGame.mockReturnValue(of(expectedGame));
    gameD2BConverter.convert
      .calledWith(expectedGame)
      .mockReturnValue(expectedBoundaryGame);

    startGameInteractor.startGame().subscribe({
      next: (game) => {
        expect(game).toStrictEqual(expectedBoundaryGame);
        done();
      },
      error: (error) => done(error),
    });
  });
});

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/StartGameInteractorTest.java
package eu.optas.use_cases.implementation;

import eu.optas.domain.Board;
import eu.optas.domain.Game;
import eu.optas.gateway.api.GameGateway;
import eu.optas.use_cases.api.BoundaryBoard;
import eu.optas.use_cases.api.BoundaryGame;
import eu.optas.utils.GameState;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class StartGameInteractorTest {
    @Test
    void startGame() {
        GameGateway gameGatewayMock = mock(GameGateway.class);
        GameD2BConverter gameD2BConverterMock = mock(GameD2BConverter.class);
        StartGameInteractor startGameInteractor = new StartGameInteractor(gameGatewayMock, gameD2BConverterMock);

        Game expectedGame = new Game(
                "123",
                GameState.IN_PROGRESS,
                mock(Board.class),
                25,
                0
        );
        BoundaryGame expectedBoundaryGame = new BoundaryGame(
                "123",
                GameState.IN_PROGRESS,
                mock(BoundaryBoard.class),
                25,
                0
        );

        when(gameGatewayMock.createGame()).thenReturn(expectedGame);
        when(gameD2BConverterMock.convert(expectedGame)).thenReturn(expectedBoundaryGame);

        BoundaryGame returnedGame = startGameInteractor.startGame();
        assertThat(returnedGame).isEqualTo(expectedBoundaryGame);
    }
}